### PR TITLE
Sort smart

### LINKED ISSUE #53: Sorting node dying
Upon sorting top hashtags, if too many hashtags exist (latest break was >100,000), the sorting node consumes too much CPU power for quite a while, and its thread is subsequently killed, killing the sorting node along with it.

While we do have key expiry, the sorting node writing back to these hashtag keys with each sort makes it mostly obsolete.

One solution may be to have the hashtag posts expire once their popularity score reaches 0 (which will happen over time, as the popularity score decays). Then, we can filter these posts out, and hashtag post lists with a length of 0 will be removed from the key space.

### PR DESCRIPTION
**Pull request by edolinsky**
Sorting node implemented to design spec

Resolves #53 (we may still run into the issue, but it's mostly a matter of tweaking from this point forward)

Adds time decay to post popularity

Smarter hashtag sorting, top sorting, trending sorting functionality has largely remained unchanged (mostly refactoring)... tests for all.

### CODE REVIEW

File: app/services/sources/ImgurSource.java [hunk 1]
```diff
@@ -103,8 +106,8 @@ public long getQueryDelta() {
 
     private Post createPost(JsonObject object) {
         Post.Builder builder = Post.newBuilder();
-        builder.setId(object.getAsJsonPrimitive("id").getAsString());
-        builder.setTimestamp(object.getAsJsonPrimitive("datetime").getAsLong());
+        builder.setId(object.getAsJsonPrimitive("id").getAsString());           // convert from seconds to millis
```
Comment (reoliveira) [inline app/services/sources/ImgurSource.java]: Is this comment meant to be on the next line?

Comment (edolinsky) [reply app/services/sources/ImgurSource.java]: Yeah, I found it a little more readable this way

File: app/services/sorting/Calculator.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    /**
+     * Calculates and inserts popularity score to each post in a list of posts
+     *
+     * @param posts list of posts
+     * @return the same list of posts, but with each post now containing a popularity score
+     */
+    public List<Post> calculatePopularityScoreOfAllPosts(List<Post> posts) {
+        List<Post> calculatedPosts = new ArrayList<>();
+
+        posts.forEach(post -> {
```
Comment (reoliveira) [inline app/services/sorting/Calculator.java]: whole method could be replaced with 
`return posts.stream().map(p -> calculatePopularityAndRebuild(post)).collect(Collectors.toList())`
 to be even simpler

Comment (reoliveira) [reply app/services/sorting/Calculator.java]: Alternatively intellij will probably suggest turning the bit inside the map function into a method reference.

File: app/services/sorting/Calculator.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+     */
+    private static int calculatePopularityScore(int numComments, int numLikes, int numShares, long timeStamp) {
+
+        // evaluate linear time decay
+        double decayFactor = (POST_EXPIRY - (System.currentTimeMillis() - timeStamp)) / (double) POST_EXPIRY;
+
+        // evaluate popularity score, scaled by linear time decay
+        double popularity = (
+
+                decayFactor * (COMMENT_WEIGHT * numComments
```
Comment (reoliveira) [inline app/services/sorting/Calculator.java:75]: Slightly odd style for java but meh.

File: app/services/sorting/PostSorter/HashtagPostSorter.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    @Override
+    public Map<String, List<Post>> sort(List<Post> posts) {
+        Map<String, List<Post>> postsByHashTag = new HashMap<>();
+
+        // sort by popularity, in decreasing order
+        List<Post> listOfPostsClone = new ArrayList<>(posts);
+        listOfPostsClone.stream()
+                .sorted(Collections.reverseOrder(Comparator.comparingInt(Post::getPopularityScore)))
+                .collect(Collectors.toList());
```
Comment (reoliveira) [inline app/services/sorting/PostSorter/HashtagPostSorter.java:35]: `.collect(Collectors.toList)` is being used but the result isn't being placed anywhere.

Also this seems to be the only place listsOfPostsClone is being used. What is its purpose?

File: app/services/sorting/PostSorter/TrendingPostSorter.java [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+
+        // retrieve old posts from specified display channel
+        List<Post> oldPosts = expandPostLists(dataSource.getAllDisplayPostLists(displayName));
+        Map<String, Post> oldPostIdMap = new HashMap<>();
+
+        // load old posts into map of unique IDs to post
+        oldPosts.forEach(post -> oldPostIdMap.put(post.getId(), post));
+
+        // calculate popularity velocity for each new post, relative to the same post in the past
+        newPosts.forEach(newPost -> {
```
Comment (reoliveira) [inline app/services/sorting/PostSorter/TrendingPostSorter.java]: Getting a bit big for an in line method, would consider extracting to a declared method.

File: app/services/sources/ImgurSource.java (hunk 1 above)
Comment (reoliveira) [reply app/services/sources/ImgurSource.java]: I disagree, I find it very confusing to have a comment on a line for which it does not pertain.

**Review by reoliveira — APPROVED** [review]
looks good

**Pull request merged**